File: cli/internal/api/measurements.go
package api

import (
	"context"
	"net/http"
	"net/url"
)

// MetricDefinition mirrors the API's metric-definition JSON — the tracked-stat
// vocabulary. Name is the slug-shaped key every command addresses it by; Label is
// the display string. Direction ∈ higher_better|lower_better records which way
// improves.
type MetricDefinition struct {
	Name      string `json:"name"`
	Label     string `json:"label"`
	Unit      string `json:"unit"`
	Direction string `json:"direction"`
	Category  string `json:"category"`
}

// MetricDefinitionCreate is the body for POST /api/v1/metrics. An omitted Label is
// derived from Name server-side.
type MetricDefinitionCreate struct {
	Name      string `json:"name"`
	Label     string `json:"label,omitempty"`
	Unit      string `json:"unit"`
	Direction string `json:"direction"`
	Category  string `json:"category"`
}

// MetricDefinitionUpdate is the body for PUT /api/v1/metrics/{name} — a partial
// update where an omitted field is left unchanged. Name is immutable.
type MetricDefinitionUpdate struct {
	Label     *string `json:"label,omitempty"`
	Unit      *string `json:"unit,omitempty"`
	Direction *string `json:"direction,omitempty"`
	Category  *string `json:"category,omitempty"`
}

// Measurement mirrors one dated reading in the API's measurement JSON.
type Measurement struct {
	Metric     string  `json:"metric"`
	MeasuredOn string  `json:"measured_on"`
	Source     string  `json:"source"`
	Notes      string  `json:"notes"`
	Value      float64 `json:"value"`
	ID         int64   `json:"id"`
}

// MeasurementCreate is the body for POST /api/v1/measurements. MeasuredOn defaults
// to today server-side when empty; Source defaults to "manual".
type MeasurementCreate struct {
	Metric     string  `json:"metric"`
	MeasuredOn string  `json:"measured_on,omitempty"`
	Notes      string  `json:"notes,omitempty"`
	Value      float64 `json:"value"`
}

// TrendPoint is one (date, value) in a metric's series, oldest-first.
type TrendPoint struct {
	MeasuredOn string  `json:"measured_on"`
	Value      float64 `json:"value"`
}

// MetricTrend mirrors GET /api/v1/metrics/{name}/trend — a metric's series plus its
// summary numbers. First/Latest/Change are nil for an unmeasured metric.
type MetricTrend struct {
	First     *float64     `json:"first"`
	Latest    *float64     `json:"latest"`
	Change    *float64     `json:"change"`
	Metric    string       `json:"metric"`
	Label     string       `json:"label"`
	Unit      string       `json:"unit"`
	Direction string       `json:"direction"`
	Category  string       `json:"category"`
	Points    []TrendPoint `json:"points"`
	Count     int          `json:"count"`
}

// KindCount, WeekCount, LibraryStats, SessionStats, Stats mirror the stats payload.
type KindCount struct {
	Kind  string `json:"kind"`
	Count int    `json:"count"`
}

type WeekCount struct {
	WeekStart string `json:"week_start"`
	Count     int    `json:"count"`
}

type LibraryStats struct {
	ByKind         []KindCount `json:"by_kind"`
	TotalMovements int         `json:"total_movements"`
	Favorites      int         `json:"favorites"`
}

type SessionStats struct {
	ByWeek     []WeekCount `json:"by_week"`
	Total      int         `json:"total"`
	Last30Days int         `json:"last_30_days"`
}

// Stats mirrors GET /api/v1/stats — the whole stats-page payload.
type Stats struct {
	Metrics  []MetricTrend `json:"metrics"`
	Library  LibraryStats  `json:"library"`
	Sessions SessionStats  `json:"sessions"`
}

// MeasurementFilter carries the list-endpoint query params.
type MeasurementFilter struct {
	Metric string
	From   string
	To     string
}

func (f MeasurementFilter) query() string {
	q := url.Values{}
	if f.Metric != "" {
		q.Set("metric", f.Metric)
	}
	if f.From != "" {
		q.Set("from", f.From)
	}
	if f.To != "" {
		q.Set("to", f.To)
	}
	if len(q) == 0 {
		return ""
	}
	return "?" + q.Encode()
}

// ListMetrics returns the metric-definition vocabulary (GET /api/v1/metrics).
func (c *Client) ListMetrics(ctx context.Context) ([]MetricDefinition, error) {
	var metrics []MetricDefinition
	if err := c.get(ctx, "/api/v1/metrics", &metrics); err != nil {
		return nil, err
	}
	return metrics, nil
}

// DefineMetric defines a metric (POST /api/v1/metrics).
func (c *Client) DefineMetric(ctx context.Context, in MetricDefinitionCreate) (MetricDefinition, error) {
	var m MetricDefinition
	if err := c.send(ctx, http.MethodPost, "/api/v1/metrics", in, &m); err != nil {
		return MetricDefinition{}, err
	}
	return m, nil
}

// UpdateMetric applies a partial update to a definition (PUT /api/v1/metrics/{name}).
func (c *Client) UpdateMetric(ctx context.Context, name string, in MetricDefinitionUpdate) (MetricDefinition, error) {
	var m MetricDefinition
	if err := c.send(ctx, http.MethodPut, "/api/v1/metrics/"+url.PathEscape(name), in, &m); err != nil {
		return MetricDefinition{}, err
	}
	return m, nil
}

// DeleteMetric removes a metric definition (DELETE /api/v1/metrics/{name}). A metric
// still referenced by a recorded measurement is a 409.
func (c *Client) DeleteMetric(ctx context.Context, name string) error {
	return c.send(ctx, http.MethodDelete, "/api/v1/metrics/"+url.PathEscape(name), nil, nil)
}

// Trend returns one metric's series + summary (GET /api/v1/metrics/{name}/trend),
// optionally windowed by from/to.
func (c *Client) Trend(ctx context.Context, metric string, f MeasurementFilter) (MetricTrend, error) {
	q := url.Values{}
	if f.From != "" {
		q.Set("from", f.From)
	}
	if f.To != "" {
		q.Set("to", f.To)
	}
	path := "/api/v1/metrics/" + url.PathEscape(metric) + "/trend"
	if len(q) > 0 {
		path += "?" + q.Encode()
	}
	var t MetricTrend
	if err := c.get(ctx, path, &t); err != nil {
		return MetricTrend{}, err
	}
	return t, nil
}

// ListMeasurements returns readings matching the filter (GET /api/v1/measurements).
func (c *Client) ListMeasurements(ctx context.Context, f MeasurementFilter) ([]Measurement, error) {
	var measurements []Measurement
	if err := c.get(ctx, "/api/v1/measurements"+f.query(), &measurements); err != nil {
		return nil, err
	}
	return measurements, nil
}

// RecordMeasurement appends a reading (POST /api/v1/measurements).
func (c *Client) RecordMeasurement(ctx context.Context, in MeasurementCreate) (Measurement, error) {
	var m Measurement
	if err := c.send(ctx, http.MethodPost, "/api/v1/measurements", in, &m); err != nil {
		return Measurement{}, err
	}
	return m, nil
}

// Stats returns the aggregated stats-page payload (GET /api/v1/stats).
func (c *Client) Stats(ctx context.Context) (Stats, error) {
	var s Stats
	if err := c.get(ctx, "/api/v1/stats", &s); err != nil {
		return Stats{}, err
	}
	return s, nil
}

File: web/src/api/measurements.ts
// Typed access to the metrics, measurements, and stats endpoints. Types mirror the
// Go API's JSON wire contract; the DB owns id, timestamps, and source defaults.
import { http } from './client'

export type MetricDirection = 'higher_better' | 'lower_better'
export type MetricCategory = 'strength' | 'cardio' | 'mobility' | 'body'

// MetricDefinition names a tracked metric and carries the metadata a chart needs:
// its unit, which direction is improvement, and a category for grouping. `name` is
// the slug-shaped key the API and CLI address it by; `label` is what the UI shows.
// `how_to_measure` is the protocol that produces the number, as markdown — empty
// means undocumented, which the about-modal says out loud rather than hiding.
export interface MetricDefinition {
  name: string
  label: string
  unit: string
  direction: MetricDirection
  category: MetricCategory
  how_to_measure: string
  created_at: string
}

// An omitted label is derived from the name server-side ("5k-time" → "5k Time").
export interface MetricDefinitionCreate {
  name: string
  label?: string
  unit: string
  direction: MetricDirection
  category: MetricCategory
  how_to_measure?: string
}

// A partial update — an omitted field is left unchanged. The name is immutable.
export interface MetricDefinitionUpdate {
  label?: string
  unit?: string
  direction?: MetricDirection
  category?: MetricCategory
  how_to_measure?: string
}

// Measurement is one dated reading of a metric — a point in the time series.
export interface Measurement {
  id: number
  metric: string
  value: number
  measured_on: string
  source: string
  notes: string
  created_at: string
}

// MeasurementCreate records a reading. measured_on defaults to today server-side;
// source defaults to "manual".
export interface MeasurementCreate {
  metric: string
  value: number
  measured_on?: string
  notes?: string
}

export interface MeasurementUpdate {
  value?: number
  measured_on?: string
  notes?: string
}

export interface TrendPoint {
  measured_on: string
  value: number
}

// MetricTrend is a metric's full series (oldest-first) plus the summary numbers a
// card shows. first/latest/change are null for a defined-but-unmeasured metric.
export interface MetricTrend {
  metric: string
  label: string
  unit: string
  direction: MetricDirection
  category: MetricCategory
  how_to_measure: string
  points: TrendPoint[]
  first: number | null
  latest: number | null
  change: number | null
  count: number
}

export interface KindCount {
  kind: string
  count: number
}

export interface WeekCount {
  week_start: string
  count: number
}

export interface LibraryStats {
  by_kind: KindCount[]
  total_movements: number
  favorites: number
}

export interface SessionStats {
  by_week: WeekCount[]
  total: number
  last_30_days: number
}

// Stats is the whole stats-page payload — one GET renders the page.
export interface Stats {
  metrics: MetricTrend[]
  library: LibraryStats
  sessions: SessionStats
}

export interface MeasurementFilter {
  metric?: string
  from?: string
  to?: string
}

function queryString(filter: MeasurementFilter): string {
  const params = new URLSearchParams()
  if (filter.metric) params.set('metric', filter.metric)
  if (filter.from) params.set('from', filter.from)
  if (filter.to) params.set('to', filter.to)
  const q = params.toString()
  return q ? `?${q}` : ''
}

export const metricsApi = {
  list: () => http.get<MetricDefinition[]>('/metrics'),
  get: (name: string) => http.get<MetricDefinition>(`/metrics/${encodeURIComponent(name)}`),
  define: (body: MetricDefinitionCreate) => http.post<MetricDefinition>('/metrics', body),
  update: (name: string, body: MetricDefinitionUpdate) => http.put<MetricDefinition>(`/metrics/${encodeURIComponent(name)}`, body),
  remove: (name: string) => http.del(`/metrics/${encodeURIComponent(name)}`),
  trend: (name: string, window: { from?: string; to?: string } = {}) => {
    const params = new URLSearchParams()
    if (window.from) params.set('from', window.from)
    if (window.to) params.set('to', window.to)
    const q = params.toString()
    return http.get<MetricTrend>(`/metrics/${encodeURIComponent(name)}/trend${q ? `?${q}` : ''}`)
  },
}

export const measurementsApi = {
  list: (filter: MeasurementFilter = {}) => http.get<Measurement[]>(`/measurements${queryString(filter)}`),
  get: (id: number) => http.get<Measurement>(`/measurements/${id}`),
  record: (body: MeasurementCreate) => http.post<Measurement>('/measurements', body),
  update: (id: number, body: MeasurementUpdate) => http.put<Measurement>(`/measurements/${id}`, body),
  remove: (id: number) => http.del(`/measurements/${id}`),
}

export const statsApi = {
  get: () => http.get<Stats>('/stats'),
}

export const CATEGORY_LABELS: Record<MetricCategory, string> = {
  strength: 'Strength',
  cardio: 'Cardio',
  mobility: 'Mobility',
  body: 'Body',
}

// slugifyMetricName derives the API's slug-shaped key from a typed display label, so
// the new-metric form can lead with the human name ("Row Machine 500m") and fill the
// key ("row-machine-500m") behind it. It is the inverse of the server's label
// derivation; the field stays editable because the key is permanent once measurements
// reference it.
export function slugifyMetricName(label: string): string {
  return label
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, '-')
    .replace(/^-+|-+$/g, '')
}

export type ChangeVerdict = 'improved' | 'worse' | 'flat'

// changeVerdict maps a trend's net change to improvement, worse, or flat — using
// direction so a smaller 5k time reads as an improvement. A "body" metric has no
// inherent good/bad direction, so its change is always neutral ("flat"). Null for an
// unmeasured metric.
export function changeVerdict(trend: MetricTrend): ChangeVerdict | null {
  if (trend.change === null) return null
  if (trend.change === 0 || trend.category === 'body') return 'flat'
  const wentUp = trend.change > 0
  const isGood = wentUp === (trend.direction === 'higher_better')
  return isGood ? 'improved' : 'worse'
}

// formatValue renders a measurement number for display. JS already drops a trailing
// ".0" (185.0 === 185) and keeps real decimals (202.5), so String() is the format —
// the named helper keeps the intent legible at call sites.
export function formatValue(value: number): string {
  return String(value)
}
